org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,165 @@
+								IPath fullPath = source.getCorrespondingResource().getProjectRelativePath().append(relative);
+								IFile newFile = resource.getProject().getFile(fullPath);
+								if (newFile.exists()) {
+									return new FileEditorInput(newFile);
+								}
+							}
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						List<IClasspathEntry> classpathEntries = new ArrayList<>();
+						for (IClasspathEntry classpathEntry : project.getRawClasspath()) {
+							if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+								classpathEntries.add(classpathEntry);
+							}
+						}
+						
+						for (IClasspathEntry sourceFolder : classpathEntries) {
+							if (sourceFolder.getOutputLocation() != null && sourceFolder.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+								IPath relative = resource.getFullPath().removeFirstSegments(sourceFolder.getOutputLocation().segmentCount());
+								Optional<IPackageFragmentRoot> source = FluentIterable.from(project.findPackageFragmentRoots(sourceFolder))
+										.first();
+								if (source.isPresent()) {
+									IPath fullPath = source.get().getCorrespondingResource().getProjectRelativePath().append(relative);
+									IFile newFile = resource.getProject().getFile(fullPath);
+									if (newFile.exists()) {
+										return new FileEditorInput(newFile);
+									}
+								}
+							}
+						}
+
+					}
+				}
+			}
+		} catch (JavaModelException e) {
+			LOG.error("Error finding the original source for the editor input", e);
+		}
+
+		return input;
+	}
+
+	public IEditorInput findOriginalSource(IEditorInput input) {
+		IFile resource = ResourceUtil.getFile(input);
+		if (resource == null) {
+			return input;
+		}
+
+		IEditorInput original = findOriginalSourceForOuputFolderCopy(input);
+		if (original != input) {
+			return original;
+		}
+
+		IEclipseTrace trace = traceInformation.getTraceToSource(resource);
+		if (trace == null) {
+			return input;
+		}
+
+		Optional<? extends ILocationInEclipseResource> sourceInformation = Iterables.tryFind(trace.getAllAssociatedLocations(),
+				(candidate) -> languageInfo.equals(candidate.getLanguage()));
+
+		if (!sourceInformation.isPresent()) {
+			return input;
+		}
+
+		IStorage originalStorage = sourceInformation.get().getPlatformResource();
+		if (originalStorage != null) {
+			return EditorUtils.createEditorInput(originalStorage);
+		}
+		
+		return null;
+	}
+
+	/**
+	 * @return the {@code IPackageFragmentRoot} that contains the provided {@code element}
+	 * @throws NullPointerException
+	 *             if the provided {@code element} is null
+	 */
+	public IPackageFragmentRoot getPackageFragmentRoot(IJavaElement element) {
+		Preconditions.checkNotNull(element, "Couldn't find source package fragment root.");
+		return element instanceof IPackageFragmentRoot ? (IPackageFragmentRoot) element : getPackageFragmentRoot(element.getParent());
```

Comment:
We probably want to use `org.eclipse.jdt.core.IJavaElement.getAncestor(int)` here

Reply:
Very good suggestion. Tx.